clang/test/SemaSYCL/unevaluated-function.cpp

```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -fcxx-exceptions -fsycl-is-device -verify -fsyntax-only %s
+// expected-no-diagnostics
+
+// Check that a function used in an unevaluated context is not subject
+// to delayed device diagnostics
+
+bool foo() {
+  // Throw exception which is not allowed on device.
+  throw 10;
+}
+
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc();
+}
+
+int main() {
+  using F1 = decltype(foo());
+  kernel_single_task<class fake_kernel>([]() {
+    using F2 = decltype(foo());
+
```

Comment:
Just wondering...
Is everything ok with the following test cases:
```
using F2 = decltype(foo()); // use foo in decltype
foo(); // then just call it on device, diagnostic is expected

sizeof(foo()); // try other unevaluated expressions

// Is everything ok with templates?
template <typename T>
T bar(T t) {
  throw t;
  return t;
}

decltype(bar(decltype(foo))); // I'd expected no diagnostics

int bar1() {
  __float128 a;
 return 1; 
}

decltype(bar1()); // Is not diagnosed too, right?

```

Reply:
They work as expected. Added them to the test, thanks.